pallene/checker.lua

```diff
@@ -391,14 +463,52 @@ function FunChecker:check_function(lambda, func_typ)
     end)
 end
 
+function FunChecker:check_and_declare_var(decl, exp)
+    local typ
+    typ, exp = self:check_initializer_exp(decl,
+        exp, "declaration of local variable %s",
+        decl.name)
+    self:add_local(decl.name, typ)
+    decl._name = self.p.symbol_table:find_symbol(decl.name)
+    return exp
+end
+
 function FunChecker:check_stat(stat)
     local tag = stat._tag
     if     tag == "ast.Stat.Decl" then
-        local typ
-        typ, stat.exp = self:check_initializer_exp(stat.decl, stat.exp,
-            "declaration of local variable %s", stat.decl.name)
-        self:add_local(stat.decl.name, typ)
-        stat.decl._name = self.p.symbol_table:find_symbol(stat.decl.name)
+        if #stat.exps == 0 then
+            for i = 1, #stat.decls do
+                local decl = stat.decls[i]
+                stat.exps[i] = self:check_and_declare_var(decl, stat.exps[i])
+            end
+        else
+            local nlast_exp = #stat.exps
+            local last_exp = stat.exps[nlast_exp]
+            for i = 1, #stat.exps do
+                local decl = stat.decls[i]
+                if decl then
+                    stat.exps[i] = self:check_and_declare_var(decl,
+                                        stat.exps[i])
+                end
+            end
+            if last_exp and last_exp._types then
+                for i = 2, #last_exp._types do
+                    local decl = stat.decls[nlast_exp + i - 1]
+                    local exp = ast.Exp.ExtraRet(last_exp.loc, last_exp, i,
+                                    #last_exp._types)
+                    exp._type = last_exp._types[i]
+                    if decl then
+                        exp = self:check_and_declare_var(decl, exp)
+                    end
+                    table.insert(stat.exps, exp)
+                end
+            end
+            if #stat.decls ~= #stat.exps then
+                type_error(stat.loc,
+                    "left-hand side expects %d value(s) but right-hand " ..
+                    "side produces %d value(s)", #stat.decls, #stat.exps)
+            end
+        end
```

Comment:
I wonder if there would be a way to implement Stat.Decl and Toplevel.Var in a similar manner. They both do basically the same thing so having different implementations smells funny.

BTW, one of the prerequisites for implementing closures in the future will be to change the Toplevel.Var to be Stat.Decls instead. Having a similar implementation for both will also help with that.

Reply:
Makes sense. Fixed.